### ISSUE TITLE
Session.run with fetches = list apparently works incorrectly

### ISSUE THREAD

**Issue by BaruchYoussin** [tensorflow/tensorflow#13133]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: The problem manifests in a standard tutorial and I wrote custom code to better understand it.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: ('v1.3.0-rc2-20-g0787eee', '1.3.0')
- **Python version**: Python 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609] on linux2
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: no GPU
- **GPU model and memory**: none

### Describe the problem
The [TensorFlow Mechanics 101 tutorial](https://tensorflow.rstudio.com/tensorflow/articles/tutorial_tensorflow_mechanics.html) code outputs value of the same Op, loss, in two ways: summary and console output.  The two values do not match.

The code is located [here](https://tensorflow.rstudio.com/tensorflow/articles/examples/mnist_fully_connected_feed.html).
To make the problem clear, I changed the console output format on line 305 to 8 decimals:

      cat(sprintf('Step %d: loss = %.8f (%.3f sec)\n',

I compared the console output for loss with the results of the summary Op (the op is defined on lines 120 and 262, and run on line 308).  I obtained the results of the summary op from the csv file downloaded from tensorboard.

The console output is close to the csv file but different, often in the third decimal; they should be identical since the op is the same on the same step.

I do not copy the output here since it is unstable, changes from one run to another.  (I have not been able to stabilize it with `use_session_with_seed` or with `tf$set_random_seed`.)

I changed lines 297-298 from

    values <- sess$run(list(train_op, loss), feed_dict = feed_dict)
    loss_value <- values[[2]]

to a supposedly equivalent code:

    sess$run(train_op, feed_dict = feed_dict)
    loss_value = sess$run(loss, feed_dict = feed_dict)
 
After that the console output agreed fully with summary data obtained from csv.

I take it that `sess$run(list(...` works incorrectly.

I do not know whether the problem is in TensorFlow or in the R interface to it.

**Comment by ppwwyyxx** [comment]
```
values <- sess$run(list(train_op, loss), feed_dict = feed_dict)
loss_value <- values[[2]]
```
is not equivalent to 
```
sess$run(train_op, feed_dict = feed_dict)
loss_value = sess$run(loss, feed_dict = feed_dict)
```

This has been discussed many times, e.g. #12863 ,#10860

**Comment by BaruchYoussin** [comment]
Thanks!

I see that if the `fetches` argument in `Session.run` is a list then its elements are executed in the random order.

I think it is worthwhile to mention this in the [manual](https://www.tensorflow.org/api_docs/python/tf/Session#run).

Thus, there is no bug.

However, I want to make this thread a suggestion to improve the above manual.

**Comment by yaroslavvb** [comment]
I'm not sure that "random order" makes sense to add to the manual because it's an implementation detail. IE, maybe today it's random order, and tomorrow it becomes left to right. Leaving it unspecified gives freedom to change implementation around without violating API guarantees

Currently the order is determined by [this part](https://github.com/tensorflow/tensorflow/blob/5747ec38cf9ffaa586b854910af8b57e148e9543/tensorflow/core/common_runtime/executor.cc#L1414) in executor.cc. It schedules ready ops simultaneously which means the order is determined by how fast computations run for each fetch

**Comment by BaruchYoussin** [comment]
I think it is important to add to the manual that the fetches are run in "unspecified order, possibly in parallel" to dispel the natural assumption that they are run in the order they appear in the list.

In this way you guys will have less threads to answer and close, and we the beginners will have to spend less time breaking our heads figuring out why our code does not work.

In fact, I remember checking that manual a few times before submitting [a question to stackoverflow](https://stackoverflow.com/questions/46212821/tensorflow-using-tfsummary-in-tensorflow-mechanics-101-tutorial/46279588#46279588) but the idea that the fetches may run not in the specified order, did not come to my mind.

**Comment by PuchatekwSzortach** [comment]
This problem has bitten my butt badly at least twice in the past. 
The worst part is that `session.run(ops)` does evaluate them in order they are provided on `tensorflow`, but not in `tensorflow-gpu`.
@yaroslavvb - A warning to user about undefined order in documentation would only be fair. If at some point order will become defined, then that can documentation can be updated. Until then you will have people scratching their heads whenever they test code along the lines of
```python
_, actual_weights = session.run([train_op, weights_op])
assert expected_weights == actual_weights
```
that works as expected with `tensorflow`, but not with `tensorflow-gpu`.

**Comment by yaroslavvb** [comment]
@PuchatekwSzortach do you want to send a PR for this? Since you are the user, you are the best positioned to find the best place to encounter this in documentation.

**Comment by PuchatekwSzortach** [comment]
@yaroslavvb Alright, I'll send a PR :)

**Comment by PuchatekwSzortach** [comment]
@yaroslavvb https://github.com/tensorflow/tensorflow/pull/17727

**Comment by Pibborn** [comment]
Sorry to post on this old issue again; however, shouldn't the information about the ```fetches``` execution order being undefined be put higher up in the documentation? [Currently](https://www.tensorflow.org/api_docs/python/tf/Session#run), it is in the "Returns" subsection, but there is an "args" subsection which seems to make more sense (to me at least).

Should I open another issue for this?

**Comment by PuchatekwSzortach** [comment]
It makes more sense to me that properties of return values are described in return section. I suppose it's best to ask others and see what's the majority's opinion.
